barretenberg/cpp/src/barretenberg/vm2/constraining/relations/range_check.test.cpp

```diff
@@ -0,0 +1,162 @@
+    uint8_t non_dynamic_bits = 32;
+    uint8_t dynamic_bits = num_bits - non_dynamic_bits;
+
+    // Choose a value that has num_bits
+    uint128_t value = (static_cast<uint128_t>(1) << num_bits) - 3;
+    uint256_t value_u256 = uint256_t::from_uint128(value);
+
+    uint16_t u16_r0 = value & 0xFFFF;
+    uint16_t u16_r1 = (value >> 16) & 0xFFFF;
+    uint16_t dynamic_slice_register = (value >> 32) & 0xFFFF;
+
+    uint16_t dynamic_bits_pow_2 = static_cast<uint16_t>(1 << dynamic_bits);
+    uint16_t dynamic_diff = static_cast<uint16_t>(dynamic_bits_pow_2 - dynamic_slice_register - 1);
+
+    TestTraceContainer trace({ {
+        { C::range_check_sel_rng_chk, 1 },
+        { C::range_check_value, value_u256 },
+        { C::range_check_rng_chk_bits, num_bits },
+        { C::range_check_is_lte_u48, 1 },
+        { C::range_check_u16_r0, u16_r0 },
+        { C::range_check_u16_r1, u16_r1 },
+        { C::range_check_u16_r7, dynamic_slice_register },
+        { C::range_check_dyn_rng_chk_bits, dynamic_bits },
+        { C::range_check_dyn_rng_chk_pow_2, dynamic_bits_pow_2 },
+        { C::range_check_dyn_diff, dynamic_diff },
+        // Enable 16 bit range lookup for each regular (non-dynamic) register being used.
+        { C::range_check_sel_r0_16_bit_rng_lookup, 1 },
+        { C::range_check_sel_r1_16_bit_rng_lookup, 1 },
+    } });
+
+    check_relation<range_check>(trace.as_rows());
+}
+
+TEST(AvmConstrainingTest, RangeCheckNegativeFull)
+{
+
+    uint8_t num_bits = 34;
+    uint8_t non_dynamic_bits = 32;
+    uint8_t dynamic_bits = num_bits - non_dynamic_bits;
+
+    // Choose a value that has num_bits
+    uint128_t value = (static_cast<uint128_t>(1) << num_bits) - 3;
+    uint256_t value_u256 = uint256_t::from_uint128(value);
+
+    uint16_t u16_r0 = value & 0xFFFF;
+    uint16_t u16_r1 = (value >> 16) & 0xFFFF;
+    uint16_t dynamic_slice_register = (value >> 32) & 0xFFFF;
+
+    uint16_t dynamic_bits_pow_2 = static_cast<uint16_t>(1 << dynamic_bits);
+    uint16_t dynamic_diff = static_cast<uint16_t>(dynamic_bits_pow_2 - dynamic_slice_register - 1);
+
+    TestTraceContainer trace({ {
+        { C::range_check_sel_rng_chk, 1 },
+        { C::range_check_value, value_u256 },
+        { C::range_check_rng_chk_bits, num_bits },
+        { C::range_check_is_lte_u48, 1 },
+        { C::range_check_u16_r0, u16_r0 },
+        { C::range_check_u16_r1, u16_r1 },
+        { C::range_check_u16_r7, dynamic_slice_register },
+        { C::range_check_dyn_rng_chk_bits, dynamic_bits },
+        { C::range_check_dyn_rng_chk_pow_2, dynamic_bits_pow_2 },
+        { C::range_check_dyn_diff, dynamic_diff },
+        // Enable 16 bit range lookup for each regular (non-dynamic) register being used.
+        { C::range_check_sel_r0_16_bit_rng_lookup, 1 },
+        { C::range_check_sel_r1_16_bit_rng_lookup, 0 }, // BAD! SHOULD BE 1
+    } });
+
+    EXPECT_THROW_WITH_MESSAGE(check_relation<range_check>(trace.as_rows()), "Relation range_check");
```

Comment:
Negative tests should probably be more granular (non full); if sth else starts failing, this statement will pass. See other negative tests. Probably you want to check a given subrelation(s).